pkg/logging/log.go

```
@@ -4,22 +4,81 @@ import (
 	"strings"
 
 	"go.uber.org/zap"
+	"go.uber.org/zap/buffer"
 	"go.uber.org/zap/zapcore"
 )
 
+// ZapConfig defines the configurable parameters you can pass in when
+// setting up the logger. See examples where we call logging.Config
+type ZapConfig struct {
+	Environment      string
+	LoggingLevel     string
+	StacktraceLength int
+}
+
+// ZapConfigOption is the type for the possible options you can pass in
+// to logging.Config
+type ZapConfigOption func(*ZapConfig)
+
+// WithEnvironment provides an option to pass in the environment
+func WithEnvironment(environment string) ZapConfigOption {
+	return func(c *ZapConfig) {
+		c.Environment = environment
+	}
+}
+
+// WithLoggingLevel provides an option to pass in the logging level
+func WithLoggingLevel(level string) ZapConfigOption {
+	return func(c *ZapConfig) {
+		c.LoggingLevel = level
+	}
+}
+
+// WithStacktraceLength provides an option to pass in the stack trace length
+func WithStacktraceLength(length int) ZapConfigOption {
+	return func(c *ZapConfig) {
+		c.StacktraceLength = length
+	}
+}
+
 // Config configures a Zap logger based on the environment string and debugLevel
-func Config(env string, loggingLevel string) (*zap.Logger, error) {
+func Config(opts ...ZapConfigOption) (*zap.Logger, error) {
+	const (
+		defaultEnvironment      = "development"
+		defaultLoggingLevel     = "info"
+		defaultStacktraceLength = 6
+	)
```

Why are these necessary, they don't seem to differ from the defaults at https://github.com/transcom/mymove/pull/5442/files#diff-51bef48edb06413d6896391f5f1cbdbb1ac7d668a387498fb151306e05b84f04R80

---

They're necessary if you don't pass in any particular option. For example, if you call `logging.Config` with only the environment and logging level, like this:

```golang
logger, err = logging.Config(
    logging.WithEnvironment(v.GetString(cli.LoggingEnvFlag)),
    logging.WithLoggingLevel(v.GetString(cli.LoggingLevelFlag)),
)
```
Then, if we don't set a default stack trace length, it will be 0, which will show the full stack trace. I could go either way on this. One option would be to imply that if you want to make sure there is a stack trace limit, then pass in the stack trace length env var, otherwise, leave it off, and you'll get the full stack trace.

Alternatively, we could say, if you leave it off, there's no harm because it will default to 6 to match the env var default value, so that we comply with STIG rules all the time. If you want to make sure it's configurable to a different value, then pass in the option.

I think I prefer the latter for safety.

---

After I typed that and though about it some more, I think I changed my mind. The whole point of the functional options is that you can leave some out if you don't need them. For example, if some files are only ever used in development, then they don't need to pass in the stack trace length, and the assumption would be that it should default to the full stack trace. Otherwise, the only way to get a bigger stack trace, is to pass in the option so that the env var can be used. OK, I will remove those defaults.

---

Okay before you were forced with the constructor though to pass all of the args right?  One thing I think I learned on Saber was that infra prefers to fail fast explicitly but maybe that's not true across the board.

I agree it's a good idea to enforce the defaults for the STIG, it's tempting to pass the viper var where those defaults already exist but it's probably not good to introduce that dependency to the log package.

---

Yeah, and even if we wanted to import the `cli` package, we couldn't because that would cause a circular dependency because `cli_test.go` imports the `logging` package. As for failing fast, that will still happen if the env var is not set to the right value, but if we forget to pass in the stack trace option, then it would default to showing the full stack trace. One way to solve that, while keeping the flexibility of passing in only the needed options, is to throw an error if the environment is *not* development, and the option has not been passed in. Or, we could just pass it in everywhere. What I wanted to avoid is that, if in the future we add some more logging options that don't apply everywhere, we would have to make changes to more files than necessary.